src/cart/src/cart/crt_context.c

```diff
@@ -1293,37 +1261,94 @@ crt_progress(crt_context_t crt_ctx, int64_t timeout,
 			hg_timeout = timeout;
 	}
 
-	while (true) {
-		crt_context_timeout_check(ctx);
-		crt_exec_progress_cb(ctx);
+	/** loop until callback returns non-null value */
+	while ((rc = cond_cb(arg)) == 0) {
+		int ret;
 
+		/**
+		 * Call progress once before processing timeouts in case
+		 * in case any replies are pending in the queue
+		 */
 		rc = crt_hg_progress(&ctx->cc_hg_ctx, hg_timeout);
-		if (rc && rc != -DER_TIMEDOUT) {
+		if (unlikely(rc && rc != -DER_TIMEDOUT))
 			D_ERROR("crt_hg_progress failed with %d\n", rc);
-			D_GOTO(out, rc = 0);
+
+		crt_context_timeout_check(ctx);
+		crt_exec_progress_cb(ctx);
+
+		ret = crt_hg_progress(&ctx->cc_hg_ctx, hg_timeout);
+		if (unlikely(ret && ret != -DER_TIMEDOUT)) {
+			D_ERROR("crt_hg_progress failed with %d\n", ret);
+			return ret;
 		}
 
-		/** execute callback */
-		rc = cond_cb(arg);
-		if (rc > 0)
-			D_GOTO(out, rc = 0);
-		if (rc < 0)
-			D_GOTO(out, rc);
+		/** check for timeout */
+		if (timeout < 0)
+			continue;
+		if (timeout == 0) {
+			/** try callback once last time just in case */
+			rc = cond_cb(arg);
+			if (unlikely(rc != 0))
+				break;
+			return -DER_TIMEDOUT;
+		}
 
-		/** check for timeout, if not infinite */
-		if (timeout > 0) {
-			now = d_timeus_secdiff(0);
-			if (now >= end) {
-				rc = -DER_TIMEDOUT;
+		now = d_timeus_secdiff(0);
+		if (now >= end) {
```

Comment:
any reason to not combine this and line 1288 together as;
if (timeout == 0 || now >= end) ?

Reply:
probably because need not take timestamp for (timeout == 0) case.

Reply:
indeed, but probably harmless at this point.